src/rulesets/Base/array.jl

```diff
@@ -1,3 +1,42 @@
+#####
+##### constructors
+#####
+
+ChainRules.@non_differentiable (::Type{T} where {T<:Array})(::UndefInitializer, args...)
+
+function rrule(::Type{T}, x::AbstractArray) where {T<:Array}
+    project_x = ProjectTo(x)
+    Array_pullback(ȳ) = (NoTangent(), project_x(ȳ))
+    return T(x), Array_pullback
+end
+
+#####
+##### `vect`
+#####
+
+@non_differentiable Base.vect()
+
+# Don't worry about projection here. The data passes straight through, so if a cotangent has
+# the wrong type for some reason, it must be the fault of another rule somewhere.
```

Comment:
```suggestion
# Fallback case -- because ProjectTo isn't safe on arbitrary types, we just assume they are correct
```

Reply:
I'm not sure that this quite captures the sentiment I was after -- this method is really for the special case associated with this method (diagonal dispatch case), as opposed to being a generic fallback: https://github.com/JuliaLang/julia/blob/bdacfa21ef7d9657be0ff3220947ceca92a340a1/base/array.jl#L126

Reply:
Oh right, sorry. I mis-read and somehow assumed `X::Vararg{T` didn't impose this. So my comment is completely wrong.

Perhaps this would help me not mess it up next time, and is shorter:
```suggestion
# Case of uniform type `T`: the data passes straight through, so no projection should be required.
```

Reply:
Sounds good.